File: packages/api/src/types/miscTypes.ts
import type { Rate } from "./blockTypes";
import type { ResponseCore } from "./commonTypes";

export type MiscBasicResponse = {
  code: number;
  data: {
    block: {
      hash: string;
      time: string;
      epoch_no: number;
      block_no: number;
      slot_no: number;
      proto: number;
    };
    ads: {
      data: {
        content: string;
        icon: string;
        section: string;
        title: string;
        type: string;
        link: string;
        text: string;
      };
      type: string;
    }[];
    version: {
      const: number;
      rate: number;
    };
    instance: {
      readonly: boolean;
      server: string;
      snapshot: string;
      time: string;
    };
    rate: Rate;
    rate_day: Rate;
  };
  tokens: number;
  ex: number;
  debug: boolean;
};

export interface BasicRate {
  date: string;
  adausd: number | null;
  btcusd: number | null;
}

export interface MiscRateResponse {
  code: number;
  data: { rates: BasicRate[] };
  tokens: number;
}

interface MiscConstResponseDataEpoch {
  no: number;
  start_time: string;
  end_time: string;
  tx_count: number;
  blk_count: number;
  fees: number;
  out_sum: number;
}

interface MiscConstResponseDataLabelData {
  scriptHash: string;
  contractAddress: string;
}

export interface MiscConstResponseDataLabel {
  name: string;
  type: string;
  label: string;
  data: MiscConstResponseDataLabelData | null;
}

interface MiscConstResponseDataEpochParam {
  nonce: string;
  epoch_no: number;
  influence: string;
  max_epoch: number;
  min_fee_a: number;
  min_fee_b: number;
  price_mem: string;
  price_step: string;
  key_deposit: number;
  max_bh_size: number;
  max_tx_size: number;
  max_val_size: number;
  pool_deposit: number;
  extra_entropy: null;
  max_tx_em_mem: number;
  min_pool_cost: number;
  max_block_size: number;
  min_utxo_size: number;
  protocol_major: number;
  protocol_minor: number;
  max_tx_ex_steps: number;
  decentralisation: number;
  max_block_ex_mem: number;
  collateral_percent: number;
  max_block_ex_steps: number;
  optimal_pool_count: number;
  coint_per_utxo_size: number;
  monetary_expand_rate: string;
  treasury_growth_rate: string;
  max_collateral_inputs: number;
}

interface MiscConstResponseDataEpochStatRewards {
  member: null;
  leader: null;
}

interface MiscConstResponseDataEpochStatStakePools {
  minting: number;
  registered: number;
}

interface MiscConstResponseDataEpochStatStake {
  epoch: number;
  active: number;
  pools: MiscConstResponseDataEpochStatStakePools;
  accounts: number;
}

interface MiscConstResponseDataEpochStatPoots {
  fees: number;
  utxo: number;
  rewards: number;
  slot_no: number;
  block_id: number;
  deposits: number;
  reserves: number;
  treasury: number;
}

interface MiscConstResponseDataEpochStatEpoch {
  start_time: string;
  end_time: string;
  block_count: number;
  tx_count: number;
  fees: number;
  out_sum: number;
  block_size: number;
}

interface MiscConstResponseDataEpochStatPoolStat {
  pools: number;
  pct_leader: null | number;
  pct_member: null | number;
  epoch_stake: null | number;
  delegator_count: null | number;
  delegator_avg: null | number;
}

interface MiscConstResponseDataEpochStatProto {
  min: number;
  max: number;
}
interface MiscConstResponseDataEpochStat {
  epoch_no: number;
  spendable_epoch: number;
  rewards: MiscConstResponseDataEpochStatRewards;
  stake: MiscConstResponseDataEpochStatStake;
  pots: MiscConstResponseDataEpochStatPoots;
  epoch: MiscConstResponseDataEpochStatEpoch;
  pool_stat: MiscConstResponseDataEpochStatPoolStat;
  proto: MiscConstResponseDataEpochStatProto;
  daily: {
    stat: {
      block_version: {
        version: number;
        count: number;
      }[];
    };
  }[];
}

export interface MiscConstResponseData {
  no: number;
  epoch: MiscConstResponseDataEpoch;
  circulating_supply: number;
  labels: MiscConstResponseDataLabel[];
  epoch_param: MiscConstResponseDataEpochParam;
  epoch_stat: MiscConstResponseDataEpochStat;
  live_stake: number | null;
}

interface MiscMarketStatus {
  low: number;
  high: number;
  open: number;
  close: number;
  volume: number;
  time_open: string;
  market_cap: number;
  time_close: string;
}

interface MiscMarketFiat {
  aud: [number, number];
  bgn: [number, number];
  brl: [number, number];
  cad: [number, number];
  chf: [number, number];
  cny: [number, number];
  czk: [number, number];
  dkk: [number, number];
  eur: [number, number];
  gbp: [number, number];
  hkd: [number, number];
  huf: [number, number];
  idr: [number, number];
  ils: [number, number];
  inr: [number, number];
  isk: [number, number];
  jpy: [number, number];
  krw: [number, number];
  mxn: [number, number];
  myr: [number, number];
  nok: [number, number];
  nzd: [number, number];
  php: [number, number];
  pln: [number, number];
  ron: [number, number];
  sek: [number, number];
  sgd: [number, number];
  thb: [number, number];
  try: [number, number];
  usd: [number, number];
  xdr: [number, number];
  zar: [number, number];
}

export interface MiscMarketData {
  date: string;
  epoch_no: number;
  need_fix: string;
  ada: MiscMarketStatus;
  btc: MiscMarketStatus;
  fiat: MiscMarketFiat;
}

interface MiscNew {
  message: {
    [key: string]: string;
  };
}

export interface MiscSearch {
  url: string;
  ident: string;
  title: string;
  category: string;
  extra: {
    icon: "balance" | "time" | "hot" | "promo";
    type: "balance" | "time" | "stake";
    value: number | string;
    id?: string;
  };
}

interface Tier {
  rq_min: number;
  rq_day: number;
  tok_day: number;
  license: string;
}

export interface MiscApiData {
  starter: Tier;
  basic: Tier;
  pro: Tier;
}

export interface Poll {
  name: string;
  url: string;
  applied: string | null;
  date_start: string;
  date_end: string;
  description: string;
  options: string[];
  state: "available" | "closed";
  vote: {
    power: number;
    is_valid: boolean;
    option: string;
    date: string;
  } | null;
  result: Record<string, { count: number; power: number }> | null;
}

export type GroupType = "pool" | "ident" | "asset" | "drep" | "collection";

interface MiscValidate {
  type: string;
  ident: string;
  valid: boolean;
}

export type MiscValidateResponse = ResponseCore<MiscValidate>;
export type PollListResponse = ResponseCore<Poll[]>;
export type MiscConstResponse = ResponseCore<MiscConstResponseData>;
export type MiscMarketResponse = ResponseCore<MiscMarketData>;
export type MiscNewResponse = ResponseCore<MiscNew>;
export type MiscSearchResponse = ResponseCore<MiscSearch[] | MiscSearch>;
export type MiscApiResponse = ResponseCore<{
  plans: MiscApiData;
}>;

File: packages/cexplorer-api-rs/src/types/misc_types.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use crate::types::common_types::ResponseCore;
use crate::types::block_types::Rate;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicBlock {
    pub hash: String,
    pub time: String,
    #[serde(default)]
    pub epoch_no: Option<f64>,
    #[serde(default)]
    pub block_no: Option<f64>,
    #[serde(default)]
    pub slot_no: Option<f64>,
    #[serde(default)]
    pub proto: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicAd {
    pub data: Value,
    #[serde(rename = "type")]
    pub ad_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicVersion {
    #[serde(rename = "const")]
    #[serde(default)]
    pub const_version: Option<f64>,
    #[serde(default)]
    pub rate: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicInstance {
    #[serde(default)]
    pub readonly: Option<bool>,
    pub server: String,
    pub snapshot: String,
    pub time: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicData {
    pub block: MiscBasicBlock,
    #[serde(default)]
    pub ads: Option<Vec<MiscBasicAd>>,
    pub version: MiscBasicVersion,
    pub instance: MiscBasicInstance,
    pub rate: Rate,
    pub rate_day: Rate,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscBasicResponse {
    #[serde(default)]
    pub code: Option<f64>,
    pub data: MiscBasicData,
    #[serde(default)]
    pub tokens: Option<f64>,
    #[serde(default)]
    pub ex: Option<f64>,
    #[serde(default)]
    pub debug: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BasicRate {
    pub date: String,
    #[serde(default)]
    pub adausd: Option<f64>,
    #[serde(default)]
    pub btcusd: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscRateData {
    pub rates: Vec<BasicRate>,
}

pub type MiscRateResponse = ResponseCore<MiscRateData>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscConstLabel {
    pub name: String,
    #[serde(rename = "type")]
    pub label_type: String,
    pub label: String,
    #[serde(default)]
    pub data: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscConstData {
    #[serde(default)]
    pub no: Option<f64>,
    #[serde(default)]
    pub epoch: Option<Value>,
    #[serde(default)]
    pub circulating_supply: Option<f64>,
    #[serde(default)]
    pub labels: Option<Vec<MiscConstLabel>>,
    #[serde(default)]
    pub epoch_param: Option<Value>,
    #[serde(default)]
    pub epoch_stat: Option<Value>,
    #[serde(default)]
    pub live_stake: Option<f64>,
}

pub type MiscConstResponse = ResponseCore<MiscConstData>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscMarketStatus {
    #[serde(default)]
    pub low: Option<f64>,
    #[serde(default)]
    pub high: Option<f64>,
    #[serde(default)]
    pub open: Option<f64>,
    #[serde(default)]
    pub close: Option<f64>,
    #[serde(default)]
    pub volume: Option<f64>,
    pub time_open: String,
    #[serde(default)]
    pub market_cap: Option<f64>,
    pub time_close: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscMarketData {
    pub date: String,
    #[serde(default)]
    pub epoch_no: Option<f64>,
    pub need_fix: String,
    pub ada: MiscMarketStatus,
    pub btc: MiscMarketStatus,
    pub fiat: Value,
}

pub type MiscMarketResponse = ResponseCore<MiscMarketData>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscSearchExtra {
    pub icon: String,
    #[serde(rename = "type")]
    pub extra_type: String,
    pub value: Value,
    #[serde(default)]
    pub id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscSearch {
    pub url: String,
    pub ident: String,
    pub title: String,
    pub category: String,
    pub extra: MiscSearchExtra,
}

pub type MiscSearchResponse = ResponseCore<Value>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Poll {
    pub name: String,
    pub url: String,
    #[serde(default)]
    pub applied: Option<String>,
    pub date_start: String,
    pub date_end: String,
    pub description: String,
    pub options: Vec<String>,
    pub state: String,
    #[serde(default)]
    pub vote: Option<Value>,
    #[serde(default)]
    pub result: Option<Value>,
}

pub type PollListResponse = ResponseCore<Vec<Poll>>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscValidate {
    #[serde(rename = "type")]
    pub validate_type: String,
    pub ident: String,
    #[serde(default)]
    pub valid: Option<bool>,
}

pub type MiscValidateResponse = ResponseCore<MiscValidate>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscApiTier {
    #[serde(default)]
    pub rq_min: Option<f64>,
    #[serde(default)]
    pub rq_day: Option<f64>,
    #[serde(default)]
    pub tok_day: Option<f64>,
    pub license: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscApiData {
    pub starter: MiscApiTier,
    pub basic: MiscApiTier,
    pub pro: MiscApiTier,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MiscApiPlans {
    pub plans: MiscApiData,
}

pub type MiscApiResponse = ResponseCore<MiscApiPlans>;
